### PR TITLE
Trainee reports entry creation/update and list

### PR DESCRIPTION
**Pull request by maonat**
Trainee reports list and modification of their records

## Description

* New `GET` /api/course-reports/{traineeId} which allows to print out all reports of a trainee having its ID provided as parameter. It will return the courseReportId and the report in a groped way:
``` example
{
  "courseReportId": 1,
  "report": [
    {
      "Quo repudiandae repellendus iure accusamus.": [
        {
          "id": 1,
          "description": "Optio ut enim dolore cum enim.",
          "done": true,
          "mastered": true
        },
        {
          "id": 2,
          "description": "Earum neque sit quaerat ducimus quia.",
          "done": true,
          "mastered": false
        },
        {
          "id": 4,
          "description": "Suscipit et nemo in dolore quas.",
          "done": true,
          "mastered": false
        }
      ],
      "Cum architecto distinctio molestias et.": [
        {
          "id": 3,
          "description": "Beatae maxime qui sapiente rem est quidem est.",
          "done": true,
          "mastered": true
        },
        {
          "id": 5,
          "description": "Reiciendis atque quia omnis non corrupti reiciendis quis.",
          "done": false,
          "mastered": false
        }
      ]
    }
  ]
}
```
* new `PUT` /api/report-entry-to-course-report that will allow to upsert a record having provided the following body:
``` body
{
  "reportEntryId": number,
  "courseReportId": number,
  "done": boolean,
  "mastered": boolean
}
```
* Generated few new tables that are required for the reports feature: Reports, ReportEntries, ReportEntryGroups, CourseReports, ReportEntryToCourseReports
* Added seeds for all new tables created

## What part of the app is affected?
* [ ] Frontend
* [x] Backend
* [ ] Infrastructure

### CODE REVIEW

File: backend/src/course-reports/course-reports.controller.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+export class CourseReportsController {
+  constructor(
+    private readonly courseReportsService: CourseReportsService,
+    private readonly reportEntriesService: ReportEntriesService,
+  ) {}
+
+  @ApiResponse({ type: '' })
+  @Get(':traineeId')
+  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+  async getTraineeReport(@Param('traineeId', ParseIntPipe) id: number) {
```
Comment (bibixx) [inline backend/src/course-reports/course-reports.controller.ts]: 1. Missing return type DTO
2. Missing response `type` for swagger (should be the same DTO as in point 1) 
3. Not needed `eslint-disable-next-line`

File: backend/src/course-reports/course-reports.controller.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      newValue.push(entry);
+
+      acc[groupDescription] = newValue;
+
+      return acc;
+    }, {});
+
+    return {
+      courseReportId: traineeReport.data.report.id,
+      report: [groupedEntries],
```
Comment (bibixx) [inline backend/src/course-reports/course-reports.controller.ts]: The returned type of the report is incorrect (ie. not like we have discussed).

Right now it is
```json
"report": [
  {
    "GROUP NAME": [
      {
        "id": 1,
        "description": "ENTRY DESCRIPTION",
        "done": true,
        "mastered": true
      },
    ]
  }
]
```

It was supposed to be
```json
"report": [
  {
    "groupDescription": "GROUP NAME",
    "entries": [
      {
        "id": 1,
        "description": "ENTRY DESCRIPTION",
        "done": true,
        "mastered": true
      },
    ]
  }
]
```

File: backend/src/course-reports/course-reports.service.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      return getFailure('REPORT_NOT_FOUND_FROM_TRAINEE');
+    }
+
+    const existAlready = await this.findOneByTraineeId(trainee.id);
+
+    if (existAlready !== null) {
+      return getFailure('REPORT_ALREADY_CREATED_FOR_TRAINEE');
+    }
+
+    const newTraineeReport = await this.courseReportsRepository.create({
```
Comment (bibixx) [inline backend/src/course-reports/course-reports.service.ts]: Redundant `await`

File: backend/src/drivers-license-category/entities/drivers-license-category.entity.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+@Entity()
+export class DriversLicenseCategory {
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @Column({ type: 'text', unique: true })
+  name: string;
+
+  @OneToMany<Report>('Report', (report) => report.driversLicenseCategory)
+  reports: Report[];
```
Comment (bibixx) [inline backend/src/drivers-license-category/entities/drivers-license-category.entity.ts]: I don't think we need this relation to be defined on this side (it should be enough to just define it in `Report`)

File: shared/src/dto/reportEntryToCourseReport/reportEntryToCourseReport.dto.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  @Type(() => CourseReportDto)
+  courseReport: CourseReportDto;
+
+  @ApiProperty()
+  @IsBoolean()
+  done: boolean;
+
+  @ApiProperty()
+  @IsBoolean()
+  mastered: boolean;
```
Comment (bibixx) [inline shared/src/dto/reportEntryToCourseReport/reportEntryToCourseReport.dto.ts]: Just an FYI, no need to add validation decorators to the response DTOs

File: backend/src/reports/reports.controller.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { Controller } from '@nestjs/common';
```
Comment (bibixx) [inline backend/src/reports/reports.controller.ts]: If we don't have anything in the reports controller then I think we could just remove it

File: backend/src/reports/reports.service.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    const reports = await this.reportsRepository.find({
+      where: {
+        driversLicenseCategory: {
+          id: driversLicenseCategoryId,
+        },
+      },
+    });
+
+    return reports;
+  }
```
Comment (bibixx) [inline backend/src/reports/reports.service.ts]: I think this method isn't being used anywhere

File: backend/src/report-entry-groups/report-entry-groups.module.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { Module } from '@nestjs/common';
```
Comment (bibixx) [inline backend/src/report-entry-groups/report-entry-groups.module.ts]: From what I can see this module isn't being used anywhere (doesn't have any controllers and the service isn't being used anywhere). I'd propse then to remove the module and the service and move the `report-entry-groups.entity.ts` to `backend/src/report-entries/entities/report-entry-groups.entity.ts`

File: backend/src/report-entry-to-course-report/entities/report-entry-to-course-report.entity.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    'CourseReport',
+    (courseReport) => courseReport.reportEntryToCourseReports,
+    {
+      nullable: false,
+    },
+  )
+  courseReport: CourseReport;
+
+  @Exclude()
+  @Column()
```
Comment (bibixx) [inline backend/src/report-entry-to-course-report/entities/report-entry-to-course-report.entity.ts]: I think we don't need to set this as a `@Column()`

File: backend/src/report-entry-to-course-report/entities/report-entry-to-course-report.entity.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    'ReportEntry',
+    (reportEntry) => reportEntry.reportEntryToCourseReports,
+    {
+      nullable: false,
+    },
+  )
+  reportEntry: ReportEntry;
+
+  @Exclude()
+  @Column()
```
Comment (bibixx) [inline backend/src/report-entry-to-course-report/entities/report-entry-to-course-report.entity.ts]: I think we don't need to set this as a `@Column()`

File: backend/src/report-entry-to-course-report/report-entry-to-course-report.controller.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  ) {}
+
+  // @ApiResponse({ type: '' })
+  // @Put()
+  // async createOrModify(
+  //   @Body()
+  //   {
+  //     reportEntryToCourseReport,
+  //   }: ReportEntryToCourseReportCreateOrModifyRequestDto,
+  // ): Promise<ReportEntryToCourseReportCreateOrModifyResponseDto>;
```
Comment (bibixx) [inline backend/src/report-entry-to-course-report/report-entry-to-course-report.controller.ts]: Commented out code

File: backend/src/report-entry-to-course-report/report-entry-to-course-report.controller.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+  @ApiResponse({
+    type: '',
+    description: 'Create or update entries for course reports',
+  })
+  @Put()
+  async createOrModify(
+    @Body()
+    reportEntryToCourseReport: ReportEntryToCourseReportCreateOrUpdateRequestDto,
+  ): Promise<any> {
```
Comment (bibixx) [inline backend/src/report-entry-to-course-report/report-entry-to-course-report.controller.ts]: 1. Missing return type DTO
2. Missing response type for swagger (should be the same DTO as in point 1)

File: backend/src/report-entry-to-course-report/report-entry-to-course-report.controller.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    await this.reportEntryToCourseReportService.setEntryForCourseReport(
```
Comment (bibixx) [inline backend/src/report-entry-to-course-report/report-entry-to-course-report.controller.ts]: This controller doesn't throw an error (probably 404) if this method returns an error

File: backend/src/report-entry-to-course-report/report-entry-to-course-report.service.ts [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+    if (
+      checkExistingEntry.data.done !== done ||
+      checkExistingEntry.data.mastered !== mastered
+    ) {
+      await this.reportEntryToCourseReportsRepository.update(
+        { id: checkExistingEntry.data.id },
+        { done, mastered },
+      );
+    }
```
Comment (bibixx) [inline backend/src/report-entry-to-course-report/report-entry-to-course-report.service.ts]: Maybe we could move this to a separate method called `upsertReportEntryToCourseReportsRepository`?

Additionally this function definitely could use transaction with [repeatable read](https://www.postgresql.org/docs/current/transaction-iso.html#:~:text=Table%C2%A013.1.%C2%A0Transaction%20Isolation%20Levels) isolation layer

```
@Transactional({
  isolationLevel: IsolationLevel.REPEATABLE_READ
  propagation: Propagation.REQUIRES_NEW,
})
```

File: backend/src/seeds/seed.reports.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    this.entities.push(report);
+    return report;
+  }
+}
+
+export const reportFactory = new ReportFactory();
+
+export const seedReports = () => {
+  Array.from({ length: 1 }).forEach(() => reportFactory.generate());
+};
```
Comment (bibixx) [inline backend/src/seeds/seed.reports.ts]:
```suggestion
export const seedReports = () => {
  reportFactory.generate()
};
```

File: backend/src/seeds/seed.ReportEntryToCourseReports.ts [hunk 16]
```diff
@@ -0,0 +1,55 @@
+import { ReportEntryToCourseReport } from '../report-entry-to-course-report/entities/report-entry-to-course-report.entity';
```
Comment (bibixx) [inline backend/src/seeds/seed.ReportEntryToCourseReports.ts]: This file doesn't follow `camelCase` convention.

**Review by bibixx — CHANGES_REQUESTED** [review]
At the end I've got 3 "global" topics
1. Let's try to merge all migrations into one file
2. Your migration on the `trainee` table doesn't work when run from `main` → If a user already exists, we don't set any default `driverLicenseCategory`. I have fixed this in this [commit](https://github.com/twoje-osk/twoje-osk/pull/78/commits/b84793211ca171a2bbc442d26ca2718ba7d42b91) (also, a bit more info in the commit's message)
3. I've changed the `courseReport` seeds a little bit. Now instead of generating one and then assigning the first trainee to it, for each trainee I generate new course report (all of this happens inside `the seed.trainees.ts`)

File: backend/src/drivers-license-category/entities/drivers-license-category.entity.ts (hunk 4 above)
Comment (maonat) [reply backend/src/drivers-license-category/entities/drivers-license-category.entity.ts]: If we don't plan in any way to use this in the FE, we can remove it.

File: backend/src/seeds/seed.ReportEntryToCourseReports.ts (hunk 16 above)
Comment (maonat) [reply backend/src/seeds/seed.ReportEntryToCourseReports.ts]: @bibixx what do you mean? You mean the dashes in the entity file?

File: backend/src/drivers-license-category/entities/drivers-license-category.entity.ts (hunk 4 above)
Comment (bibixx) [reply backend/src/drivers-license-category/entities/drivers-license-category.entity.ts]: No, we don't use it. We can remove this

File: backend/src/seeds/seed.ReportEntryToCourseReports.ts (hunk 16 above)
Comment (bibixx) [reply backend/src/seeds/seed.ReportEntryToCourseReports.ts]: The file should be named `seed.reportEntryToCourseReports.ts`

File: backend/src/drivers-license-category/entities/drivers-license-category.entity.ts (hunk 4 above)
Comment (maonat) [reply backend/src/drivers-license-category/entities/drivers-license-category.entity.ts]: I'm having troubles removing it from the entity. Wouldn't be enough to add the decorator Exclude()? @bibixx

**Review by bibixx — APPROVED** [review]

**Pull request merged**